src/ui/public/saved_objects/saved_objects_client.ts

```diff
@@ -0,0 +1,358 @@
+import { kfetch, KFetchQuery } from '../kfetch';
+import { keysToCamelCaseShallow, keysToSnakeCaseShallow } from '../utils/case_conversion';
+import { SavedObject } from './saved_object';
+
+interface RequestParams {
+  method: 'POST' | 'GET' | 'PUT' | 'DELETE';
+  path: string;
+  query?: KFetchQuery;
+  body?: object;
+}
+
+interface CreateOptions {
+  id?: string;
+  overwrite?: boolean;
+  migrationVersion?: MigrationVersion;
+  references?: SavedObjectReference[];
+}
+
+interface BulkCreateOptions<T extends SavedObjectAttributes = any> extends CreateOptions {
+  type: string;
+  attributes: T;
+}
+
+interface UpdateOptions {
+  version?: string;
+  migrationVersion?: MigrationVersion;
+  references?: SavedObjectReference[];
+}
+
+interface BatchResponse<T extends SavedObjectAttributes = any> {
+  savedObjects: Array<SavedObject<T>>;
+}
+
+interface FindResults<T extends SavedObjectAttributes = any> extends BatchResponse<T> {
+  total: number;
+  perPage: number;
+  page: number;
+}
+
+interface BatchQueueEntry {
+  type: string;
+  id: string;
+  resolve: <T extends SavedObjectAttributes>(value: SavedObject<T> | PlainSavedObject<T>) => void;
+  reject: (reason?: any) => void;
+}
+
+const join = (...uriComponents: Array<string | undefined>) =>
+  uriComponents
+    .filter((comp): comp is string => Boolean(comp))
+    .map(encodeURIComponent)
+    .join('/');
+
+/**
+ * Interval that requests are batched for
+ * @type {integer}
+ */
+const BATCH_INTERVAL = 100;
+
+const API_BASE_URL = '/api/saved_objects/';
+
+export class SavedObjectsClient {
+  /**
+   * Throttled processing of get requests into bulk requests at 100ms interval
+   */
+  private processBatchQueue = throttle(
+    () => {
+      const queue = cloneDeep(this.batchQueue);
+      this.batchQueue = [];
+
+      this.bulkGet(queue)
+        .then(({ savedObjects }) => {
+          queue.forEach(queueItem => {
+            const foundObject = savedObjects.find(savedObject => {
+              return savedObject.id === queueItem.id && savedObject.type === queueItem.type;
+            });
+
+            if (!foundObject) {
+              return queueItem.resolve(this.createSavedObject(pick(queueItem, ['id', 'type'])));
+            }
+
+            queueItem.resolve(foundObject);
+          });
+        })
+        .catch(err => {
+          queue.forEach(queueItem => {
+            queueItem.reject(err);
+          });
+        });
+    },
+    BATCH_INTERVAL,
+    { leading: false }
+  );
+
+  private batchQueue: BatchQueueEntry[];
+
+  constructor() {
+    this.batchQueue = [];
+  }
+
+  /**
+   * Persists an object
+   *
+   * @param {string} type
+   * @param {object} [attributes={}]
+   * @param {object} [options={}]
+   * @property {string} [options.id] - force id on creation, not recommended
+   * @property {boolean} [options.overwrite=false]
+   * @property {object} [options.migrationVersion]
+   * @returns {promise} - SavedObject({ id, type, version, attributes })
```

Comment:
How about we refer to the actualy types in order to prevent divergence with the implementation? Are these annotations worth keeping at all? They are not adding much beyond the type annotations.

Reply:
Yeah I will remove those for now, since it doesn't make much sense, all editors (and tools should too) pull out the type information from the actual typings and not the JSDoc. So unless there isn't more description that doesn't help much.